src/ssl_stubs.c

```diff
@@ -290,120 +290,92 @@ static int protocol_flags[] = {
 #endif
 };
 
-static const SSL_METHOD *get_method(int protocol, int type)
+static const SSL_METHOD *get_method(int type)
 {
   const SSL_METHOD *method = NULL;
 
   caml_release_runtime_system();
-  switch (protocol)
+  switch (type)
   {
     case 0:
-      switch (type)
-      {
-        case 0:
-          method = SSLv23_client_method();
-          break;
+      method = TLS_client_method();
+      break;
 
-        case 1:
-          method = SSLv23_server_method();
-          break;
+    case 1:
+      method = TLS_server_method();
+      break;
 
-        case 2:
-          method = SSLv23_method();
-          break;
-      }
+    case 2:
+      method = TLS_method();
       break;
+  }
+  caml_leave_blocking_section();
 
-#ifndef OPENSSL_NO_SSL3
-    case 1:
-      switch (type)
-      {
-        case 0:
-          method = SSLv3_client_method();
-          break;
+  if (method == NULL)
+    caml_raise_constant(*caml_named_value("ssl_exn_method_error"));
 
-        case 1:
-          method = SSLv3_server_method();
-          break;
+  return method;
+}
 
-        case 2:
-          method = SSLv3_method();
-          break;
+static void set_protocol(SSL_CTX *ssl_context, int protocol)
+{
+#ifdef HAVE_TLS13
+  int max_proto = TLS1_3_VERSION;
+#else
+  int max_proto = TLS1_2_VERSION;
+#endif
+  switch (protocol)
+  {
+    case 0:
+      if (!SSL_CTX_set_min_proto_version(ssl_context, SSL3_VERSION) ||
```

Comment:
Why can we remove the call to SSL_CTX_set_min_proto_version ?

Reply:
Sorry, I read the diff in the wrong direction. So no question in fact.

Reply:
I was also a bit surprised reading the diff, because it looks like:

- if the protocol is `SSLv23`, we set the max version to TLS 1.3 and min version to SSL3
- in all other cases, we set the version strictly to the protocol that has been passed in.

This sounds wrong, to say the least, and makes it impossible to support an SSL context that can e.g. communicate in either TLS v1.2 and v1.3.

I'm thinking about non-breaking ways of addressing this. Ideally we'd be able to pass min-version and max-version to create_context.

Reply:
I'm tempted to conform to the behavior described in https://www.openssl.org/docs/man3.0/man3/TLSv1_2_server_method.html#TLS_method-TLS_server_method-TLS_client_method

In OCaml-SSL, we would treat `protocol` as the min-supported version, and optionally allow passing in a max-version. What do you think?

Reply:
We could probably even break the API since we're already doing a breaking change by dropping OpenSSL v1.0.2. Something to keep in mind.

Reply:
And... thinking about it further, the best option is probably to expose bindings to `SSL_CTX_set_{min,max}_proto_version`, which is both backwards compatible and more rigorous from a binding library perspective.

Which makes this PR correct.